### PR TITLE
erc-3156 ready for review

### PR DESCRIPTION
**Pull request by albertocuestacanada**
ERC-3156 is ready for review. [Link for discussion](https://ethereum-magicians.org/t/erc-3156-flash-loans-review-discussion/5077).

The reference implementations and the lender wrappers have all been deployed to [Kovan](https://github.com/albertocuestacanada/ERC3156). Feel free as well to kick the tires using the FlashBorrower at 0xeeb0c120bF35fB0793b1c7d0D93230e552020398 contract which has a flashBorrow function to initiate flash loans on any ERC-3156 contract, and the necessary callback for at least paying the loan fees and successfully finalizing the loan.

### CODE REVIEW

**Comment by eip-automerger** [comment]
Hi! I'm a bot, and I wanted to automerge your PR, but couldn't because of the following issue(s):

 - Trying to change EIP 3156 state from Draft to Review

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
> If a fee is charged, the contract implementing flashLoan MAY use it in any desired way (e.g. the fee can be burned or transferred to any other party).

This is unnecessary in the spec.  A spec shouldn't try to define all of the things that it doesn't specify, it should focus exclusively on the things it does specify.  This makes the specification much smaller and thus easier for a human to work with and implement.

> If the flashLoan used tokens generated by a mint, they SHOULD be the target of a burn before the end of the transaction.

This feels out of scope.  This specification is defining an interface, and shouldn't try to dictate what people do inside implementations of that interface.

> The flashLoan function MUST verify that the tokens lended were returned, and MUST NOT take them from the receiver.

Rephrase this to specify what the expected behavior is if the lender has not received the lent tokens + fees.  Also note that currently this doesn't specify the fee needs to be paid here.

> The lender contract MAY mint the tokens lended, instead of executing a transfer of tokens it holds.

As above, this is out of scope of the specification and not actually a specification but rather a suggestion.  Suggestions aren't appropriate for technical specifications and they add unnecessary length that doesn't add value to interoperability.

> A receiver of flash mints MUST implement an onFlashLoan callback:

I think you mean `flash loans` here, not `flash mints`.

> For the transaction to not revert, receiver MUST send amount + fee of token to the caller.

This should be more technically specific.  Something like "amount + fee of token **MUST** be transferred to the `msg.sender` of the `onFlashLoan` function at some point prior to returning from `onFlashLoan`.

> flashSupply returns zero on unsupported contracts to allow for borrowers to discover the lending services offered by an ERC-3156 compliant lender.

This cannot be guaranteed.  You should use something like ERC-165 or ERC-1820 if you want such guarantees.

----

Are there any security issues with reentrancy that implementers need to worry about?  I thought there were some problems with naive implementations of flash loans where someone could drain the lender via reentrancy.  If so, those should be discussed in the Security Considerations section.

----

Remove the test cases section if you don't plan on inlining the tests.  External links from EIPs should be avoided as much as possible, and this section isn't required.

----

Remove the first 3 paragraphs in `## Implementation` section.  EIPs are not the appropriate venue to collect/maintain external links to things.  The inline implementation is fine, and if people want to learn more they can search the internet or check the `discussions-to` link.

**Comment by albertocuestacanada** [comment]
That's great feedback, @MicahZoltu.

 - Most items I've just accepted.
 - The specification has been rewritten more precisely.
 - There is a new section on Security Considerations, relevant to the recent DefiSaver exploit.

I don't know of any flash mint implementations with reentrancy issues. The FlashMinter implementation is naive and supports reentrancy:
```
    function flashLoan(address receiver, address token, uint256 amount, bytes calldata data) external override {
        require(token == address(this), "FlashMinter: unsupported loan currency");
        uint256 _fee = _flashFee(token, amount);
        _mint(receiver, amount);
        IERC3156FlashBorrower(receiver).onFlashLoan(msg.sender, token, amount, _fee, data);
        _burn(address(this), amount.add(_fee));
    }
```

**Comment by rekmarks** [comment]
This needs a rebase.

**Comment by MicahZoltu** [comment]
@albertocuestacanada I recommend checking out the readme (and potentially the contracts) of this repository: https://github.com/Austin-Williams/safer-flash-loans for details on the reentrancy vectors.  I think Austin also posted a blog article (or maybe it was OpenZeppelin) on the subject.

**Comment by MicahZoltu** [comment]
Mention me after this has been rebased and I'll take another look at it.

**Comment by albertocuestacanada** [comment]
I have to confess, that I have no idea how to rebase this.

**Comment by albertocuestacanada** [comment]
@MicahZoltu, I've managed to rebase this thing. Hopefully we will squash those commits on merging.

I've reached out to Austin to discuss, not only reentrancy vulnerabilities, but also his statement in that repo that not using `transferFrom` makes the lender vulnerable.

File: EIPS/eip-3156.md [hunk 1]
```diff
@@ -31,12 +31,12 @@ A flash lending feature integrates two smart contracts using a callback pattern.
 
 ### Lender Specification
 
-A `lender` smart contract implementing a flash lending feature MUST implement a `flashSupply` function:
+A `lender` smart contract implementing a flash lending feature MUST implement a `maxFlashAmount` function:
 ```
-function flashSupply(address token) external external returns (uint256);
+function maxFlashAmount(address token) external external returns (uint256);
```
Comment (fubuloubu) [inline EIPS/eip-3156.md:36]:
```suggestion
function maxFlashAmount(address token) external returns (uint256);
```

Comment (fubuloubu) [reply EIPS/eip-3156.md:36]: A couple others like this

File: EIPS/eip-3156.md [hunk 2]
```diff
@@ -55,38 +55,32 @@ function flashLoan(address receiver, address token, uint256 amount, bytes callda
...
 
 The interfaces described in this ERC have been chosen as to cover the known flash lending use cases, while allowing for safe and gas efficient implementations.
 
-`flashSupply(address token)`
-
-`flashSupply` returns zero on unsupported contracts to allow for borrowers to discover the lending services offered by an ERC-3156 compliant lender.
-
 `flashFee(address token, uint256 amount)`
 
 `flashFee` reverts on unsupported contracts, because returning a numerical value would be incorrect.
```
Comment (fubuloubu) [inline EIPS/eip-3156.md:86]:
```suggestion
`flashFee` reverts on unsupported tokens, because returning a numerical value would be incorrect.
```

**Comment by albertocuestacanada** [comment]
Upon a discussion with Austin Williams, I'm going to change the spec to have the lender pull the loan and the fee from the receiver. That will impact all sections

File: EIPS/eip-3156.md [hunk 3]
```diff
@@ -275,53 +266,41 @@ contract FlashLender is IERC3156FlashLender {
...
 
 ## Security Considerations
 
+
 ### Verification of callback arguments
 
-The arguments on the `onFlashLoan` callback can be divided in two groups, that require different checks before they can be trusted to be genuine.
+The arguments of `onFlashLoan` are expected to reflect the conditions of the flash loan, but cannot be trusted unconditionally. They can be divided in two groups, that require different checks before they can be trusted to be genuine.
 
 0. No arguments can be assumed to be genuine without some kind of verification. `sender`, `token` and `value` refer to a past transaction that might not have happened if the caller of `onFlashLoan` decides to lie. `fee` might be false or calculated incorrectly. `calldata` is not expected to have been verified or manipulated by the caller.
```
Comment (albertocuestacanada) [inline EIPS/eip-3156.md:296]:
```suggestion
0. No arguments can be assumed to be genuine without some kind of verification. `sender`, `token` and `value` refer to a past transaction that might not have happened if the caller of `onFlashLoan` decides to lie. `fee` might be false or calculated incorrectly. `data` might have been manipulated by the caller.
```

**Comment by albertocuestacanada** [comment]
@MicahZoltu, this is ready for review and discussion again. Thanks for your patience.

File: EIPS/eip-3156.md [hunk 4]
```diff
@@ -55,51 +55,43 @@ function flashLoan(address receiver, address token, uint256 amount, bytes callda
...
-If a fee is charged, the contract implementing `flashLoan` MAY use it in any desired way (e.g. the fee can be burned or transferred to any other party).
+The `lender` MUST take the `amount + fee` `token` from the `receiver`, or revert if this is not succesful.
 
 ### Receiver Specification
 
-A `receiver` of flash mints MUST implement an `onFlashLoan` callback:
+A `receiver` of flash loans MUST implement an `onFlashLoan` callback:
 
 ```
 interface FlashBorrowerLike {
```
Comment (albertocuestacanada) [inline EIPS/eip-3156.md:71]:
```suggestion
interface IERC3156FlashBorrower {
```

File: EIPS/eip-3156.md [hunk 5]
```diff
@@ -255,12 +240,18 @@ contract FlashLender is IERC3156FlashLender {
...
+     * @param data A data parameter to be passed on to the `receiver` for any custom use.
      */
-    function flashSupply(address token) external view override returns (uint256) {
-        return supportedTokens[token] ? IERC20(token).balanceOf(address(this)) : 0;
+    function flashLoan(address receiver, address token, uint256 amount, bytes calldata data) external override {
+        require(supportedTokens[token], "FlashLender: Unsupported currency");
+        uint256 fee = _flashFee(token, amount);
+        IERC20(token).transfer(receiver, amount);
+        IERC3156FlashBorrower(receiver).onFlashLoan(msg.sender, token, amount, fee, data);
+        IERC20(token).transferFrom(receiver, address(this), amount.add(fee));
```
Comment (MicahZoltu) [inline EIPS/eip-3156.md:254]: `transferFrom` doesn't necessarily throw on failure.  Sadly, some tokens don't return `true` on success either.  ERC20 spec says you **MUST** return `true` on success, so if you want to be spec compliant, wrap this call in a `require(...)`.  Same is true for the `.transfer(...)` call above.

Note: Will do a full review tomorrow, just wanted to call this one out early so you can get a head start on fixing it.

Comment (MicahZoltu) [reply EIPS/eip-3156.md:254]: Alternatively, check balance before and after the call to make sure that the transfers went through successfully.  Downside is that it is another bunch of gas to do that.

Comment (albertocuestacanada) [reply EIPS/eip-3156.md]: Fixed.

**Comment by albertocuestacanada** [comment]
Any idea what does the CI failure mean?

**Review by MicahZoltu — APPROVED** [review]
Consider adding support for flash lending ETH by making `token == SENTINEL` mean ETH in all functions.

**Comment by MicahZoltu** [comment]
The spelling mistakes were what was blocking CI, once those are fixed we can merge this.  The rest are suggestions.

File: EIPS/eip-3156.md [hunk 6]
```diff
@@ -31,12 +31,26 @@ A flash lending feature integrates two smart contracts using a callback pattern.
...
+interface IERC3156FlashLender {
+    function maxFlashAmount(address token) external view returns (uint256);
+    function flashFee(address token, uint256 amount) external view returns (uint256);
+    function flashLoan(
+        address receiver,
+        address token,
+        uint256 amount,
+        bytes calldata data
+    ) external;
+}
```
[the remark this answers is not in the document]
Comment (albertocuestacanada) [reply EIPS/eip-3156.md]: I've seen often as a style to define functions in a single line when the number of arguments is three or fewer, and expand them to an argument per line with four or more.

I've gone for each argument in a line in function definitions.

File: EIPS/eip-3156.md [hunk 7]
```diff
@@ -109,60 +178,52 @@ A `sender` will often be required in the `onFlashLoan` function, which the lende
...
-Other implementations include [WETH10](https://github.com/WETH10/WETH10) and [MakerDAO MIP-25](https://github.com/hexonaut/dss-flash/pull/18).
-
 ### Flash Borrower Reference Implementation
 
 ```
-// SPDX-License-Identifier: GPL-3.0-or-later
 pragma solidity ^0.7.5;
 
 import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import "@openzeppelin/contracts/math/SafeMath.sol";
```
[the remark this answers is not in the document]
Comment (albertocuestacanada) [reply EIPS/eip-3156.md]: Fixed.

File: EIPS/eip-3156.md [hunk 8]
```diff
@@ -46,60 +60,115 @@ function flashFee(address token, uint256 amount) external external returns (uint
...
 
-A `receiver` of flash mints MUST implement an `onFlashLoan` callback:
+A `receiver` of flash loans MUST implement an `onFlashLoan` callback:
 
 ```
-interface FlashBorrowerLike {
-    function onFlashLoan(address sender, address token, uint256 amount, uint256 fee, bytes calldata) external;
+interface IERC3156FlashBorrower {
+    function onFlashLoan(
+        address sender,
```
[the remark this answers is not in the document]
Comment (albertocuestacanada) [reply EIPS/eip-3156.md]: I also refactored from `address` to `IERC20` wherever possible, did I go too far?

File: EIPS/eip-3156.md [hunk 9]
```diff
@@ -218,23 +281,22 @@ contract ERC20FlashMinter is ERC20, IERC3156FlashLender {
     function _flashFee(address token, uint256 amount) internal view returns (uint256) {
         return fee == type(uint256).max ? 0 : amount.div(fee);
```
[the remark this answers is not in the document]
Comment (albertocuestacanada) [reply EIPS/eip-3156.md]: Fixed into being defined as bps (multiples of 0.0001 %). I was being too smart.

**Pull request merged**